source/adapters/cuda/enqueue.cpp

```diff
@@ -154,33 +153,18 @@ void guessLocalWorkSize(ur_device_handle_t Device, size_t *ThreadsPerBlock,
     GlobalSizeNormalized[i] = GlobalWorkSize[i];
   }
 
+  size_t MaxBlockDim[3];
+  MaxBlockDim[0] = MaxThreadsPerBlock[0];
   MaxBlockDim[1] = Device->getMaxBlockDimY();
   MaxBlockDim[2] = Device->getMaxBlockDimZ();
 
-  UR_CHECK_ERROR(
-      cuOccupancyMaxPotentialBlockSize(&MinGrid, &MaxBlockSize, Kernel->get(),
-                                       NULL, LocalSize, MaxThreadsPerBlock[0]));
-
-  ThreadsPerBlock[2] = std::min(GlobalSizeNormalized[2], MaxBlockDim[2]);
-  ThreadsPerBlock[1] =
-      std::min(GlobalSizeNormalized[1],
-               std::min(MaxBlockSize / ThreadsPerBlock[2], MaxBlockDim[1]));
-  MaxBlockDim[0] = MaxBlockSize / (ThreadsPerBlock[1] * ThreadsPerBlock[2]);
-  ThreadsPerBlock[0] = std::min(
-      MaxThreadsPerBlock[0], std::min(GlobalSizeNormalized[0], MaxBlockDim[0]));
-
-  static auto IsPowerOf2 = [](size_t Value) -> bool {
-    return Value && !(Value & (Value - 1));
-  };
-
-  // Find a local work group size that is a divisor of the global
-  // work group size to produce uniform work groups.
-  // Additionally, for best compute utilisation, the local size has
-  // to be a power of two.
-  while (0u != (GlobalSizeNormalized[0] % ThreadsPerBlock[0]) ||
-         !IsPowerOf2(ThreadsPerBlock[0])) {
-    --ThreadsPerBlock[0];
-  }
+  int MinGrid, MaxBlockSize;
+  UR_CHECK_ERROR(cuOccupancyMaxPotentialBlockSize(
```

Comment:
What happens here if a kernel submits a very small range e.g. 100 and has no shared memory limitations and e.g. MaxThreadsPerBlock[0] is 1024: Won't this function return a value higher than 100? Will this then be passed to the kernel execution such that it uses a range of e.g. 1024 instead of 100?

Reply:
No the `roundToHighestFactorOfGlobalSizeIn3d` will make sure that no dims exceed the global dims in any dimension

Reply:
Do you understand these failures? https://github.com/intel/llvm/actions/runs/8162721434/job/22337470532
This is the only pr that seems to affect those tests from that batch: they use parallel_for with range.

Reply:
There were some bugs in the original implementation but they weren't being tested. This patch fixes the bugs and I have added some tests here https://github.com/intel/llvm/pull/12715/files#diff-0ceed90561c114d8a66c1fdfe72fdb97ac3d3af519bffabc57e62514fe117de2

But it looks like the tests failing are all just timeouts to me. I don't know what could have changed to make this happen other than hardware/CI environment related changes

Reply:
You were right there were bugs in this causing an infinite loop :stuck_out_tongue_winking_eye: Have fixed